File: widgets/Thumbnail.py
from send2trash import send2trash

from PyQt6.QtCore import QSize, Qt, pyqtSignal
from PyQt6.QtGui import QPixmap, QIcon, QAction, QCursor
from PyQt6.QtWidgets import QToolButton, QMenu, QWidget
from functions.getDominantColour import getDominantColour


class Thumbnail(QToolButton):

    context_menu: QMenu

    deleted = pyqtSignal(QWidget)

    max_side_size: int = 1024

    origin_path: str
    pixmap: QPixmap
    source_size: QSize
    dominant_colour: int | None = None

    def __init__(self, parent, path, width):
        super().__init__(parent)

        self.pixmap = QPixmap(path)
        self.source_size = self.pixmap.size()
        if self.pixmap.width() > self.pixmap.height():
            if self.pixmap.width() > self.max_side_size:
                self.pixmap = self.pixmap.scaledToWidth(self.max_side_size)
        else:
            if self.pixmap.height() > self.max_side_size:
                self.pixmap = self.pixmap.scaledToHeight(self.max_side_size)
        self.setScaledIcon(width)
        self.origin_path = path

        self._createContextMenu()

    def _createContextMenu(self):
        self.setContextMenuPolicy(Qt.ContextMenuPolicy.CustomContextMenu)

        self.context_menu = QMenu(parent=self, title='Контекстное меню')

        delete_action = QAction(parent=self, text='Удалить файл')
        delete_action.triggered.connect(self._delete)
        self.context_menu.addAction(delete_action)

        self.customContextMenuRequested.connect(self.open_context_menu)

    def open_context_menu(self):
        self.context_menu.exec(QCursor.pos())

    def _delete(self):
        send2trash(self.origin_path)
        self.deleted.emit(self)

    def setScaledIcon(self, width):
        scaled_pixmap = self.pixmap.scaledToWidth(width)
        size = QSize(width, scaled_pixmap.height())
        self.setFixedSize(size)
        self.setIconSize(size)
        self.setIcon(QIcon(scaled_pixmap))

    def determineDominantColour(self):
        if self.dominant_colour is None:
            self.dominant_colour = getDominantColour(self.pixmap)

File: widgets/Thumbnail.cpp
#include <QFile>

#include "Thumbnail.h"

/**
 * @brief setupSlots
 *      connects the slots to window's widgets and signals
 */
void Thumbnail::setupSlots()
{
    QObject::connect(
        this, SIGNAL(customContextMenuRequested(QPoint)),
        this, SLOT(customContextMenu_Requested(QPoint))
    );
}

/**
 * @brief createCustomContextMenu
 *      define and configure the context menu
 */
void Thumbnail::createCustomContextMenu()
{
    if (this->customContextMenu != nullptr) return;

    this->customContextMenu = new QMenu(this);
    this->customContextMenuActions.clear();

    QAction *deleteAction = new QAction(tr("menu-action.delete"), this);
    QObject::connect(
        deleteAction, SIGNAL(triggered()),
        this, SLOT(deleteAction_Triggered())
    );
    this->customContextMenuActions.push_back(deleteAction);

    for (QAction *action : this->customContextMenuActions)
    {
        this->customContextMenu->addAction(action);
    }
}

/**
 * @brief deleteCustomContextMenu
 *      delete the context menu
 */
void Thumbnail::deleteCustomContextMenu()
{
    for (QAction *action : this->customContextMenuActions)
    {
        delete action;
    }

    delete this->customContextMenu;
    this->customContextMenu = nullptr;
}

/**
 * @brief load
 *      loads image into pixmap
 */
void Thumbnail::load()
{
    QPixmap pixmap(this->imagePath);
    this->srcSize = new QSize(pixmap.size());

    bool needResize = this->srcSize->width() > MAX_SIDE_SIZE || this->srcSize->height() > MAX_SIDE_SIZE;

    if (this->srcSize->width() > this->srcSize->height())
    {
        pixmap = needResize ? QPixmap(pixmap.scaledToWidth(MAX_SIDE_SIZE)) : pixmap;
    }
    else
    {
        pixmap = needResize ? QPixmap(pixmap.scaledToHeight(MAX_SIDE_SIZE)) : pixmap;
    }

    this->setIcon(QIcon(pixmap));
}

/**
 * @brief resizeToWidth
 *      resize tool button keep image ratio according to width
 */
void Thumbnail::resizeToWidth(int width)
{
    int height = this->srcSize->height() * width / this->srcSize->width();
    int srcHeight = this->srcSize->height();

    if (height > srcHeight) height = srcHeight;

    QSize newSize(
        width,
        height < MAX_SIDE_SIZE
            ? height
            : MAX_SIDE_SIZE
    );

    this->setFixedSize(newSize);
    this->setIconSize(newSize);
}

/**
 * @brief getImagePath
 *      returns source image path
 * @return
 */
QString Thumbnail::getImagePath()
{
    return this->imagePath;
}

/**
 * @brief getSrcSizeAsString
 *      returns source file resolution
 * @return
 */
QString Thumbnail::getResolution()
{
    return QString::fromStdString(to_string(this->srcSize->width()) + "x" + to_string(this->srcSize->height()));
}

/**
 * @brief deleteSourceFile
 *      deletes source file from disk
 */
void Thumbnail::deleteSourceFile()
{
    QFile::moveToTrash(this->imagePath);

    emit this->deleted();
}

/**
 * @brief customContextMenu_Requested
 *      handle signal when context menu requested
 * @param point
 */
void Thumbnail::customContextMenu_Requested(QPoint point)
{
    this->createCustomContextMenu();
    this->customContextMenu->exec(this->mapToGlobal(point));
    this->deleteCustomContextMenu();
}

/**
 * @brief deleteAction_Triggered
 *      handle when delete action is triggered
 */
void Thumbnail::deleteAction_Triggered()
{
    this->deleteSourceFile();
}

/**
 * @brief Thumbnail
 *      Constructor of the widget
 * @param parent
 *      parent widget for this one
 */
Thumbnail::Thumbnail(QWidget *parent, QString path) : QToolButton(parent)
{
    this->imagePath = path;

    this->setupSlots();
    this->setContextMenuPolicy(Qt::CustomContextMenu);
}

/**
 * Destructor of the widget
 */
Thumbnail::~Thumbnail()
{
    delete this->srcSize;
}
